plugins/src/webrtcsink/imp.rs

```diff
@@ -787,81 +759,187 @@ impl CongestionController {
+        control_op: CongestionControlOp,
+        controller_type: ControllerType,
+    ) {
+        gst_trace!(
+            CAT,
+            obj: element,
+            "consumer {}: applying congestion control operation {:?}",
+            self.peer_id,
+            control_op
+        );
 
-            match control_op {
-                CongestionControlOp::Hold => (),
-                CongestionControlOp::Increase(IncreaseType::Additive(value)) => {
-                    self.clamp_bitrate(self.target_bitrate + value as i32, n_encoders);
-                }
-                CongestionControlOp::Increase(IncreaseType::Multiplicative(factor)) => {
-                    self.clamp_bitrate((self.target_bitrate as f64 * factor) as i32, n_encoders);
-                }
-                CongestionControlOp::Decrease(factor) => {
-                    self.clamp_bitrate((self.target_bitrate as f64 * factor) as i32, n_encoders);
-
-                    // Smoothing factor
-                    let alpha = 0.75;
-                    if let Some(ema) = self.bitrate_ema {
-                        let sigma: f64 = (self.target_bitrate as f64) - ema;
-                        self.bitrate_ema = Some(ema + (alpha * sigma));
-                        self.bitrate_emvar =
-                            (1. - alpha) * (self.bitrate_emvar + alpha * sigma.powi(2));
-                    } else {
-                        self.bitrate_ema = Some(self.target_bitrate as f64);
-                        self.bitrate_emvar = 0.;
-                    }
-                }
+        let n_encoders = encoders.len() as i32;
+        let prev_bitrate = i32::min(self.target_bitrate_on_delay, self.target_bitrate_on_loss);
+        match &control_op {
+            CongestionControlOp::Hold => {}
+            CongestionControlOp::Increase(IncreaseType::Additive(value)) => {
+                self.clamp_bitrate(
+                    self.target_bitrate_on_delay + *value as i32,
+                    n_encoders,
+                    controller_type,
+                );
             }
+            CongestionControlOp::Increase(IncreaseType::Multiplicative(factor)) => {
+                self.clamp_bitrate(
+                    (self.target_bitrate_on_delay as f64 * factor) as i32,
+                    n_encoders,
+                    controller_type,
+                );
+            }
+            CongestionControlOp::Decrease { factor, .. } => {
+                self.clamp_bitrate(
+                    (self.target_bitrate_on_delay as f64 * factor) as i32,
+                    n_encoders,
+                    controller_type,
+                );
 
-            let target_bitrate = self.target_bitrate / n_encoders;
-
-            let fec_ratio = {
-                if target_bitrate <= 2000000 || self.max_bitrate <= 2000000 {
-                    0f64
+                // Smoothing factor
+                let alpha = 0.75;
+                if let Some(ema) = self.bitrate_ema {
```

Comment:
I'm a bit confused, I thought part of the benefits of this patch was that the loss-based controller would *not* update the detected congestion window (if any)? Shouldn't we only update `bitrate_ema` and `bitrate_emvar` when the controller_type is Delay ?